### PR TITLE
Discuss ISpan/IReadableSpan

### PR DESCRIPTION
**Pull request by lmolkova**
**TL;DR:**
1. should we introduce ReadableSpan as a contract for exporters?
2. should we keep ISpan/IReadbleSpan as interfaces (or make them abstract classes)
3. should we make SpanSdk implement both interfaces or separate implementation and introduce efficient conversion?

### Motivation

We need two different views for span: for adapters (`ISpan`) and for exporters. Today we use `Span` implementation as a contract for exporters. This introduces a way for processors/exporters to **abuse setters on the Span**. 

It also **couples exporters with specific Span implementation and behavior** and it might be harder to change `Span` impl in future.

Another issue is naming. [Span<T>](https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netcore-3.1) is something quite different and **we've been strongly advised by .NET team against reusing `Span` in public API surface**.

### Proposal

This change introduces `IReadableSpan` as a contract between SDK and exporters. 

*Pros*:  allows to evolve\change `Span` impl without breaking exporters and limits the API abuse possibilities

*Cons*:  requires to keep `ISpan` or  `IReadableSpan` as interfaces (multiple inheritance) which makes them not possible to be changed in any way. This is also inconsistent with the rest of APIs here.

Alternative is to have abstract classes and implement two different Spans implementing each base class  with efficient conversion between them (this basically means one extra allocation for sampled-in/exported spans).

Another alternative is to mix getters and setters in `ISpan` (and make it a contract for exporters) which goes against OTel idea of API isolation, but at the same time seems to only cause minor problems related to API abuse.

### CODE REVIEW

File: src/OpenTelemetry/Trace/Export/IReadableSpan.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry.Trace.Export
+{
+    public interface IReadableSpan
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Trace/Export/IReadableSpan.cs]: I think this would be clearer if named `IReadOnlySpan`.

Comment (lmolkova) [reply src/OpenTelemetry/Trace/Export/IReadableSpan.cs]: It would conflict with [ReadOnlySpan](https://docs.microsoft.com/en-us/dotnet/api/system.readonlyspan-1?view=netcore-3.1) 

Also ReadableSpan term was used in spec and used in [java](https://github.com/open-telemetry/opentelemetry-java/blob/023f3fef2502f270d17c9e3c6c128d92d378ec90/sdk/src/main/java/io/opentelemetry/sdk/trace/ReadableSpan.java) and might be in other languages

Comment (MikeGoldsmith) [reply src/OpenTelemetry/Trace/Export/IReadableSpan.cs]: `IReadOnlySpan` (interface) won't conflict with `ReadOnlySpan` (System.Memory). I do see there is going to be confusion between OpenTelemetry and System.Memory but I don't think we can avoid all of it.

My preference is still `IReadOnlySpan`. I don't think *Readable* means anything where *ReadOnly* is a more generally accepted concept.

@SergeyKanzhelev @cijothomas @bruno-garcia what do you think?

Comment (bruno-garcia) [reply src/OpenTelemetry/Trace/Export/IReadableSpan.cs]: Agree `ReadableSpan` isn't a great name and it's true that `IReadOnlySpan` wouldn't conflict with `ReadOnlySpan` (`I` in there) even though I got confused for a sec when I saw the name mentioned here. 

That said, `IReadOnlySpan` would be more idiomatic for .NET devs albeit confusing a bit with _the other ReadOnlySpan_. There's the cost of being foreign to people familiar with OTel (assuming `ReadableSpan` gets picked up by most other SDKs).

That said, I have no strong opinions here but just for the sake of giving one: 
In this case, it could be worth sticking with the name that one might encounter while reading up on OTel more generally, so `IReadableSpan`.

Comment (lmolkova) [reply src/OpenTelemetry/Trace/Export/IReadableSpan.cs]: done a bit of research and noticed that several OTel SDKs use `SpanData` except for those which use `Span` impl directly and do not have any Readable/SpanData notion. `ReadableSpan` seems to be used by Java only.

I believe `ISpanData` may be a good alternative as it is something common across SDKs and does not conflict with very common .NET `ReadOnlySpan`. Event `ISpan` is a problem (but seems like renaming it has a lot of drawbacks) but introducing yet another conflict seems unreasonable.

@MikeGoldsmith @SergeyKanzhelev @bruno-garcia ?

Comment (SergeyKanzhelev) [reply src/OpenTelemetry/Trace/Export/IReadableSpan.cs]: SpanData is a good name

Comment (MikeGoldsmith) [reply src/OpenTelemetry/Trace/Export/IReadableSpan.cs]: 👍 for `SpanData`

**Comment by tarekgh** [comment]
a couple of thoughts here: 

- I believe using interfaces here is not a good choice. Going with the interfaces will restrict us in the future to extend these interfaces. I think most likely the Span will need to extend in the future for any new scenarios we didn't think about now. At that time will require exposing a new interface and possibly exposing new APIs to handle the new interfaces. That for sure will complicate OT libraries and will start creating confusion. I strongly recommend using abstract classes for the Span to allow future flexibility. 

- For the read-only span, did we consider other options instead of exposing a new interface? here are some other options:
    - Exposing a simple struct that wraps the read-write span object and this struct exposes only the getters and not allowing mutating the span. The benefit here is we'll be dealing one thing (which is the Span) and not requiring the implementation classes to implement 2 interfaces. Of course, the exporter's APIs will use this struct and we never pass spans to such APIs. The Span interface can have a method like Span.GetReadOnlyInstance which returns this struct and this method will be called before sending the Spans to the exporters. There will not be any memory or speed perf overhead with this idea and will simplify our APIs.
    - The other idea is in the implementation of the Span when ending/stoping the span, mark it as read-only and don't allow any mutation after this point. We will have to ensure all spans sent to the exporters has to be stopped/ended.

**Comment by MikeGoldsmith** [comment]
@tarekgh I'm unsure why an interface is less flexible than an abstract class, please can you elaborate on this? If we were to go with an abstract class we'd still be restricted by all implementations being forced to provide abstract members & properties.

Would your suggestion of wrapping `Span` with a struct look something like this?

```csharp
public struct SpanData
{
    private readonly Span span;

    public SpanData(Span span)
    {
        this.span = span;
    }

    public string Name => this.span.Name;
    .. other properties
}
```

**Comment by lmolkova** [comment]
@tarekgh 

> Exposing a simple struct that wraps the read-write span object

This is a good alternative. The only problem I see, this struct needs to be created and populated for each sampled in span twice (onStart/OnStop). This is probably not a big deal as long as we agree that it can get a reference (not copy) to the lists of links, attributes, etc

> The other idea is in the implementation of the Span when ending/stoping the span, mark it as read-only 

I believe the idea of separation is to provide different views from API perspective, create a contract for exporters and prevent not only abuse but also confusion (who calls what).

@MikeGoldsmith interfaces could not be extended in future versions - this is a breaking change. Default interface method implementation is only possible with .NET core 3.0 and we really want to support .NET framework.
Abstract classes on other side, could have default method implementation, i.e. could be extended.

**Comment by lmolkova** [comment]
I'll experiment with SpanData struct. assuming it goes well, I'll change ISpan to abstract class and then we'll have to make the worst choice for ISpan name: `Span` is really a bad idea according to the .NET team because of the conflict with `Span<T>`. At the same time, Span is the key term in tracing world. 

.NET team advised us to use `TelemetrySpan` instead. I really hope to hear OTel community thoughts on this. I'll create a separate issue to discuss.

**Comment by tarekgh** [comment]
@MikeGoldsmith your sample code of the struct is correct. 

> This is probably not a big deal as long as we agree that it can get a reference (not copy) to the lists of links, attributes, etc

The struct should only contain the span object instance which means the size of the whole struct will be IntPtr.Size. that means it doesn't matter if you pass it by reference or by value.

**Comment by MikeGoldsmith** [comment]
Adding members to interfaces is not a breaking change. Renaming members / parameters, changing parameter / return types, etc are breaking changes both for interfaces and abstract classes. Please can you elaborate on what changes an abstract class allows that an interface could not?

A good list of breaking changes at API, binary and source levels are described here: https://stackoverflow.com/questions/1456785/a-definitive-guide-to-api-breaking-changes-in-net.

I think `TelemetrySpan` is a more accurate description but will be a clear deviation from OTel spec and I highly doubt the spec will be changed this late in development. I'd be okay with replacing `Span` but we need to do it before beta.

**Comment by lmolkova** [comment]
@MikeGoldsmith 

- imagine OTel API v1.1 adds MethodA to ISpan.
- some library calss ISpan.MethodA()
- User uses OTel SDK v1

It causes runtime Missing method exception. this is binary breaking change https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/breaking-changes.

**Comment by MikeGoldsmith** [comment]
Okay, I see the potential conflict. I was assuming the SDK would provide the API as a dependency into a user's application code, but they could reference the API directly and upgrade to an unsupported version for a given SDK.

Is that likely and/or supported? Should you just use the API version referenced by the SDK you want to use?

**Comment by lmolkova** [comment]
In this example some additional library that uses API 1.1 and user wants SDK 1.0. 

User App
  - Lib1 -> OTel API .1.1
  - OTel SDK 1.0 -> API 1.0. E.g. SDK requires API v >= 1.0

As a result, API 1.1 is chosen by .NET Core (and Framework will chose whatever is configured through binding redirects which additional pain)

We can require exactly the same version of API on the SDK, then build will fail and user will be forced to upgrade to 1.1. This seems to complicate dependency management for users and prevent any separate fixes in API. I also believe it's extremely hard to guarantee the same version assuming SDK injection in runtime.

I'd rather fallback to default .NET guideline: it's a breaking change that requires major version update and we should use abstract classes.

Actually there is exactly the same problem in Java and perhaps other languages. So any changes to ISpan will require major version update.

**Comment by tarekgh** [comment]
I felt https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/ is useful blog post so I decided to share it.

**Comment by lmolkova** [comment]
@tarekgh thanks for sharing. this is never coming to .NET framework, correct?

**Comment by bruno-garcia** [comment]
Yeah sadly even when we enable C# 8 we won’t be able to use that unless it’s ns2.1 or nca3.0+

**Comment by tarekgh** [comment]
> this is never coming to .NET framework, correct?

Yes. In https://devblogs.microsoft.com/dotnet/building-c-8-0/ section "Platform dependencies" is explicitly saying:

```
default interface member implementations rely on new runtime enhancements, and we will not make those in the .NET Runtime 4.8 either.
```

**Comment by MikeGoldsmith** [comment]
With the above points clearer, I also agree a move to using abstract classes is the best solution for managing Span :+1:

**Comment by lmolkova** [comment]
Thanks everyone, I've done `SpanData`. The minor complication is that it's not straightforward to mock it and it affects exporters testing a bit (you have to create a real span using public ISpan APIs).

We should consider adding some APIs for testing. In the meantime, I suggest to tolerate that for now (it's easy to add APIs, hard to remove)

I created a separate issue to discuss `Span` name https://github.com/open-telemetry/opentelemetry-dotnet/issues/429 and to track conversion to abstract class. This is ready for final review
